File: platforms/swift/Sources/ShopifyCheckoutKit/CheckoutDiagnostics.swift
import Foundation

/// An SDK diagnostic that applications may observe for integration telemetry.
///
/// Diagnostics are informational and never replace checkout lifecycle, preload,
/// or protocol events. Applications may safely ignore every diagnostic event.
public enum CheckoutDiagnosticEvent: Equatable, Sendable {
    /// An incoming checkout message was denied before protocol dispatch.
    case messageRejected(CheckoutMessageRejection)
}

/// Details about an incoming checkout message denied by the ingress policy.
///
/// The raw message body is intentionally omitted because rejected input is
/// untrusted and may contain sensitive or arbitrarily large data.
public struct CheckoutMessageRejection: Equatable, Sendable {
    /// The origin the message was received from, for example `https://example.com`.
    public let origin: String

    /// The stable reason the message was denied.
    public let reason: Reason

    package init(origin: String, reason: Reason) {
        self.origin = origin
        self.reason = reason
    }

    public enum Reason: Equatable, Sendable {
        /// The message was sent from a child frame rather than the checkout's main frame.
        case childFrame

        /// The message origin used explicit port zero while origin validation was enabled.
        case unsupportedPort

        /// The message origin did not match the effective allowlist.
        case originNotAllowed
    }
}

/// SDK-wide diagnostic events emitted by Checkout Kit.
///
/// Subscriptions are hot and do not replay earlier events. Subscribe before
/// calling `preload(checkout:)` when preload diagnostics are required. Listeners
/// are called on the main actor so diagnostics follow the same observation model
/// as checkout preload state.
public final class CheckoutDiagnostics: Sendable {
    /// A retained observation of SDK diagnostic events.
    ///
    /// Keep the subscription for as long as diagnostics should be observed.
    /// Observation stops when the subscription is cancelled or released.
    @MainActor
    public final class Subscription {
        private var listener: (@MainActor (CheckoutDiagnosticEvent) -> Void)?

        fileprivate init(listener: @escaping @MainActor (CheckoutDiagnosticEvent) -> Void) {
            self.listener = listener
        }

        /// Stops this listener from receiving future diagnostic events.
        public func cancel() {
            listener = nil
        }

        fileprivate func receive(_ event: CheckoutDiagnosticEvent) {
            listener?(event)
        }
    }

    @MainActor
    private final class WeakSubscription {
        weak var value: Subscription?

        init(_ value: Subscription) {
            self.value = value
        }
    }

    @MainActor private var subscriptions = [UUID: WeakSubscription]()

    package init() {}

    /// Subscribes a listener to future diagnostic events.
    ///
    /// Retain the returned subscription for as long as events should be observed.
    @MainActor
    public func subscribe(
        _ listener: @escaping @MainActor (CheckoutDiagnosticEvent) -> Void
    ) -> Subscription {
        subscriptions = subscriptions.filter { $0.value.value != nil }

        let subscription = Subscription(listener: listener)
        subscriptions[UUID()] = WeakSubscription(subscription)
        return subscription
    }

    @MainActor
    package func emit(_ event: CheckoutDiagnosticEvent) {
        log(event)

        // Retain a snapshot for this delivery so listeners may cancel themselves
        // or release other subscriptions without mutating the traversed collection.
        let currentSubscriptions = subscriptions.compactMap { $0.value.value }
        subscriptions = subscriptions.filter { $0.value.value != nil }
        for subscription in currentSubscriptions {
            subscription.receive(event)
        }
    }

    private func log(_ event: CheckoutDiagnosticEvent) {
        switch event {
        case let .messageRejected(rejection):
            OSLogger.shared.debug(
                "Rejected checkout message from \(rejection.origin): \(rejection.reason.logDescription)"
            )
        }
    }
}

extension CheckoutMessageRejection.Reason {
    fileprivate var logDescription: String {
        switch self {
        case .childFrame:
            return "message was sent from a child frame"
        case .unsupportedPort:
            return "origin uses unsupported port 0"
        case .originNotAllowed:
            return "origin is not in the allowlist"
        }
    }
}

File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/CheckoutDiagnostics.kt
package com.shopify.checkoutkit

import java.util.concurrent.CopyOnWriteArraySet
import java.util.concurrent.atomic.AtomicBoolean

/** An SDK diagnostic that applications may observe for integration telemetry. */
public sealed interface CheckoutDiagnosticEvent {
    /** An incoming checkout message was denied before protocol dispatch. */
    public data class MessageRejected(
        public val rejection: CheckoutMessageRejection,
    ) : CheckoutDiagnosticEvent
}

/**
 * Details about an incoming checkout message denied by the ingress policy.
 *
 * The raw message body is intentionally omitted because rejected input is untrusted and may
 * contain sensitive or arbitrarily large data.
 */
public data class CheckoutMessageRejection(
    /** Origin the message was received from, for example `https://example.com`. */
    public val origin: String,

    /** Stable reason the message was denied. */
    public val reason: Reason,
) {
    public enum class Reason {
        /** The message was sent from a child frame rather than the checkout's main frame. */
        CHILD_FRAME,

        /** The message origin used explicit port zero while origin validation was enabled. */
        UNSUPPORTED_PORT,

        /** The message origin did not match the effective allowlist. */
        ORIGIN_NOT_ALLOWED,
    }
}

/**
 * SDK-wide diagnostic events emitted by Checkout Kit.
 *
 * Subscriptions are hot and do not replay earlier events. Subscribe before calling `preload` when
 * preload diagnostics are required. Listener failures are isolated so diagnostics cannot interrupt
 * checkout processing or prevent other listeners from receiving an event.
 */
public class CheckoutDiagnostics internal constructor() {
    /** Receives diagnostic events on the Android main thread. */
    public fun interface Listener {
        public fun onDiagnosticEvent(event: CheckoutDiagnosticEvent)
    }

    /** A cancellable diagnostic subscription. */
    public interface Subscription : AutoCloseable {
        /** Stops this listener from receiving future events. */
        public fun cancel()

        /** Equivalent to [cancel], enabling Java try-with-resources usage. */
        override fun close(): Unit = cancel()
    }

    private val listeners = CopyOnWriteArraySet<Listener>()

    /**
     * Subscribes [listener] to future diagnostics.
     *
     * The returned subscription retains the listener until it is cancelled.
     */
    public fun subscribe(listener: Listener): Subscription {
        listeners.add(listener)
        return ListenerSubscription(listeners, listener)
    }

    internal fun emit(event: CheckoutDiagnosticEvent) {
        log(event)

        // Snapshot before dispatch so a listener added after emission cannot observe an earlier
        // event, and listeners may cancel themselves without mutating the traversed collection.
        val currentListeners = listeners.toList()
        onMainThread {
            currentListeners.forEach { listener ->
                try {
                    listener.onDiagnosticEvent(event)
                } catch (error: Exception) {
                    ShopifyCheckoutKit.log.e(DIAGNOSTICS_LOG_TAG, "Diagnostic listener threw", error)
                }
            }
        }
    }

    private fun log(event: CheckoutDiagnosticEvent) {
        when (event) {
            is CheckoutDiagnosticEvent.MessageRejected -> {
                val rejection = event.rejection
                ShopifyCheckoutKit.log.d(
                    DIAGNOSTICS_LOG_TAG,
                    "Rejected checkout message from ${rejection.origin}: ${rejection.reason.logDescription}",
                )
            }
        }
    }

    private class ListenerSubscription(
        private val listeners: CopyOnWriteArraySet<Listener>,
        private val listener: Listener,
    ) : Subscription {
        private val cancelled = AtomicBoolean(false)

        override fun cancel() {
            if (cancelled.compareAndSet(false, true)) {
                listeners.remove(listener)
            }
        }
    }

    private companion object {
        private const val DIAGNOSTICS_LOG_TAG = "CheckoutDiagnostics"
    }
}

private val CheckoutMessageRejection.Reason.logDescription: String
    get() = when (this) {
        CheckoutMessageRejection.Reason.CHILD_FRAME -> "message was sent from a child frame"
        CheckoutMessageRejection.Reason.UNSUPPORTED_PORT -> "origin uses unsupported port 0"
        CheckoutMessageRejection.Reason.ORIGIN_NOT_ALLOWED -> "origin is not in the allowlist"
    }
